bundles/org.eclipse.equinox.p2.publisher.eclipse/src/org/eclipse/equinox/p2/publisher/eclipse/BundlesAction.java

```diff
@@ -428,17 +428,30 @@ protected void addCapability(List<IProvidedCapability> caps, GenericDescription
 		capAttrs.compute(capNs,
 				(k, v) -> (v instanceof String) ? v : String.format("%s_%s-%s", iu.getId(), iu.getVersion(), capNo)); //$NON-NLS-1$
 
+		for (Version version : getVersions(capAttrs)) {
```

Comment:
Something seems wrong with the approach to have multiple capabilities represented here instead of a single multiple versioned capability.

By design the OSGi specification defined the `osgi.ee` namespace to be multi-versioned.  This is so you could specify a filter like (&(osgi.ee=JavaSE)(version=1.8)(!(version=18)).  That would be a bundle that wants to run on Java 8 - 17 but not on Java 7 or Java 18.  If you split these out into individual capabilities the more complex filters will not work.  Perhaps that is not possible anyway with p2 requirements.  But it is possible with OSGi `Require-Capability`.

Reply:
> Something seems wrong with the approach to have multiple capabilities represented here instead of a single multiple versioned capability.

As far as I can oversee it `ProvidedCapability` expects the the value of `PROPERTY_VERSION` to be a single Version object, which is for example enforced in the constructor `ProvidedCapability(String namespace, Map<String, Object> props)` and relied on in `ProvidedCapability.getVersion()`.

> 
> By design the OSGi specification defined the `osgi.ee` namespace to be multi-versioned. This is so you could specify a filter like (&(osgi.ee=JavaSE)(version=1.8)(!(version=18)). That would be a bundle that wants to run on Java 8 - 17 but not on Java 7 or Java 18. If you split these out into individual capabilities the more complex filters will not work. Perhaps that is not possible anyway with p2 requirements. But it is possible with OSGi `Require-Capability`.

So if I run a Java-18 JDK there should be one capability providing `osgi.ee=JavaSE` that has a List of versions from 1.0 up to 18? And then the suggested filter should fail to match that capability because it has one version that matches the negated version constraint? And if we would provide one capability for each version the filter would match all but the version 18 capability? So consequently the bundle would be selected/wired although it should not be selected because at least one capability is selected and the rejected one is ignored?

Changing `IProvidedCapability.getVersion()` to return a `List<Version>` would be a greater effort that likely would have to wait until the next release then. It should be possible because `getVersion()` is annotated with `@noreference`.

Reply:
It looks like `getVersion()` is directly only used in `MetadataWriter` and `org.eclipse.equinox.internal.p2.touchpoint.eclipse.Util` and tests. But there seems to be more indirect usage using obtaining the `PROPERTY_VERSION` value of the properties map.

Reply:
We have came to an agreement to move forward with this solution despite it having some limitations for more complicated matching filters that expect multiple values for the version of the capability.